// brighterscript/dist/parser/tests/Parser.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EOF = exports.rangeMatch = exports.identifier = exports.token = void 0;
const lexer_1 = require("../../lexer");
const creators_1 = require("../../astUtils/creators");
/* A set of utilities to be used while writing tests for the BRS parser. */
/**
 * Creates a token with the given `kind` and (optional) `literal` value.
 */
function token(kind, text) {
    return {
        kind: kind,
        text: text,
        isReserved: lexer_1.ReservedWords.has((text || '').toLowerCase()),
        range: creators_1.interpolatedRange,
        leadingWhitespace: ''
    };
}
exports.token = token;
/**
 * Creates an Identifier token with the given `text`.
 */
function identifier(text) {
    return token(lexer_1.TokenKind.Identifier, text);
}
exports.identifier = identifier;
/**
 * Test whether a range matches a group of elements with a `range`
 */
function rangeMatch(range, elements) {
    return range.start.line === elements[0].range.start.line &&
        range.start.character === elements[0].range.start.character &&
        range.end.line === elements[elements.length - 1].range.end.line &&
        range.end.character === elements[elements.length - 1].range.end.character;
}
exports.rangeMatch = rangeMatch;
/** An end-of-file token. */
exports.EOF = token(lexer_1.TokenKind.Eof, '\0');
