// @rokucommunity/bslint/dist/plugins/trackCodeFlow/varTracking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.runDeferredValidation = exports.createVarLinter = exports.resetVarContext = exports.VarLintError = void 0;
const brighterscript_1 = require("brighterscript");
var VarLintError;
(function (VarLintError) {
    VarLintError["UninitializedVar"] = "LINT1001";
    VarLintError["UnsafeIteratorVar"] = "LINT1002";
    VarLintError["UnsafeInitialization"] = "LINT1003";
    VarLintError["CaseMismatch"] = "LINT1004";
    VarLintError["UnusedVariable"] = "LINT1005";
})(VarLintError = exports.VarLintError || (exports.VarLintError = {}));
var ValidationKind;
(function (ValidationKind) {
    ValidationKind["Assignment"] = "Assignment";
    ValidationKind["UninitializedVar"] = "UninitializedVar";
    ValidationKind["UninitialisedFn"] = "UninitialisedFn";
    ValidationKind["Unsafe"] = "Unsafe";
})(ValidationKind || (ValidationKind = {}));
const deferredValidation = new Map();
function getDeferred(file) {
    return deferredValidation.get(file.pathAbsolute);
}
function resetVarContext(file) {
    deferredValidation.set(file.pathAbsolute, []);
}
exports.resetVarContext = resetVarContext;
function createVarLinter(lintContext, file, fun, state, diagnostics) {
    const { severity } = lintContext;
    const deferred = getDeferred(file);
    const args = new Map();
    args.set('m', { name: 'm', range: brighterscript_1.Range.create(0, 0, 0, 0), isParam: true, isUnsafe: false, isUsed: true });
    fun.parameters.forEach((p) => {
        const name = p.name.text;
        args.set(name.toLowerCase(), { name: name, range: p.name.range, isParam: true, isUnsafe: false, isUsed: false });
    });
    if (brighterscript_1.isClassMethodStatement(fun.functionStatement)) {
        args.set('super', { name: 'super', range: null, isParam: true, isUnsafe: false, isUsed: true });
    }
    function verifyVarCasing(curr, name) {
        if (curr && curr.name !== name.text) {
            diagnostics.push({
                severity: severity.caseSensitivity,
                code: VarLintError.CaseMismatch,
                message: `Variable '${name.text}' was previously set with a different casing as '${curr.name}'`,
                range: name.range,
                file: file,
                data: {
                    name: curr.name,
                    range: name.range
                }
            });
        }
    }
    function setLocal(parent, name, isIterator) {
        const key = name.text.toLowerCase();
        const arg = args.get(key);
        const local = {
            name: name.text,
            range: name.range,
            parent: parent,
            isIterator: isIterator,
            metBranches: 1,
            isUnsafe: false,
            isUsed: false
        };
        if (arg) {
            verifyVarCasing(arg, name);
            return local;
        }
        if (!parent.locals) {
            parent.locals = new Map();
        }
        else {
            verifyVarCasing(parent.locals.get(key), name);
        }
        parent.locals.set(key, local);
        deferred.push({
            kind: ValidationKind.Assignment,
            name: name.text,
            local: local,
            range: name.range
        });
        return local;
    }
    function findLocal(name) {
        var _a, _b;
        const key = name.toLowerCase();
        const arg = args.get(key);
        if (arg) {
            return arg;
        }
        const { parent, blocks, stack } = state;
        if ((_a = parent === null || parent === void 0 ? void 0 : parent.locals) === null || _a === void 0 ? void 0 : _a.has(key)) {
            return parent.locals.get(key);
        }
        for (let i = stack.length - 2; i >= 0; i--) {
            const block = blocks.get(stack[i]);
            const local = (_b = block === null || block === void 0 ? void 0 : block.locals) === null || _b === void 0 ? void 0 : _b.get(key);
            if (local) {
                return local;
            }
        }
        return undefined;
    }
    // A local was found but it is considered unsafe (e.g. set in an if branch)
    // Found out whether a parent has this variable set safely
    function findSafeLocal(name) {
        var _a;
        const key = name.toLowerCase();
        const { blocks, stack } = state;
        if (stack.length < 2) {
            return undefined;
        }
        for (let i = stack.length - 2; i >= 0; i--) {
            const block = blocks.get(stack[i]);
            const local = (_a = block === null || block === void 0 ? void 0 : block.locals) === null || _a === void 0 ? void 0 : _a.get(key);
            // if partial, look up higher in the scope for a non-partial
            if (local && !local.isUnsafe) {
                return local;
            }
        }
    }
    function openBlock(block) {
        var _a;
        const { stat } = block;
        if (brighterscript_1.isForStatement(stat)) {
            // for iterator will be declared by the next assignement statement
        }
        else if (brighterscript_1.isForEachStatement(stat)) {
            // declare `for each` iterator variable
            setLocal(block, stat.item, true);
        }
        else if ((_a = state.parent) === null || _a === void 0 ? void 0 : _a.narrows) {
            narrowBlock(block);
        }
    }
    function narrowBlock(block) {
        var _a;
        const { parent } = state;
        const { stat } = block;
        if (brighterscript_1.isIfStatement(stat) && brighterscript_1.isIfStatement(parent.stat)) {
            block.narrows = parent === null || parent === void 0 ? void 0 : parent.narrows;
            return;
        }
        (_a = parent === null || parent === void 0 ? void 0 : parent.narrows) === null || _a === void 0 ? void 0 : _a.forEach(narrow => {
            if (narrow.block === stat) {
                setLocal(block, narrow, false).narrowed = narrow;
            }
            else {
                // opposite narrowing for other branches
                setLocal(block, narrow, false).narrowed = Object.assign(Object.assign({}, narrow), { type: narrow.type === 'invalid' ? 'valid' : 'invalid' });
            }
        });
    }
    function visitStatement(curr) {
        const { stat } = curr;
        if (brighterscript_1.isAssignmentStatement(stat) && state.parent) {
            // value = stat.value;
            setLocal(state.parent, stat.name, brighterscript_1.isForStatement(state.parent.stat));
        }
    }
    function closeBlock(closed) {
        const { locals, branches, returns } = closed;
        const { parent } = state;
        if (!locals || !parent) {
            if (locals) {
                finalize(locals);
            }
            return;
        }
        // when closing a branched statement, evaluate vars with partial branches covered
        if (branches > 1) {
            locals.forEach((local) => {
                if (local.metBranches !== branches) {
                    local.isUnsafe = true;
                }
                local.metBranches = 1;
            });
        }
        else if (brighterscript_1.isIfStatement(parent.stat)) {
            locals.forEach(local => {
                // keep narrowed vars if we `return` the invalid branch
                if (local.narrowed) {
                    if (!returns || local.narrowed.type === 'valid') {
                        locals.delete(local.name.toLowerCase());
                    }
                }
            });
        }
        // move locals to parent
        if (!parent.locals) {
            parent.locals = locals;
        }
        else {
            const isParentIf = brighterscript_1.isIfStatement(parent.stat);
            const isLoop = brighterscript_1.isForStatement(closed.stat) || brighterscript_1.isForEachStatement(closed.stat) || brighterscript_1.isWhileStatement(closed.stat);
            locals.forEach((local, name) => {
                var _a;
                const parentLocal = parent.locals.get(name);
                // if var is an iterator var, flag as partial
                if (local.isIterator) {
                    local.isUnsafe = true;
                }
                // if a parent var isn't partial then the var stays non-partial
                if (isParentIf) {
                    if (parentLocal) {
                        local.isUnsafe = parentLocal.isUnsafe || local.isUnsafe;
                        local.metBranches = ((_a = parentLocal.metBranches) !== null && _a !== void 0 ? _a : 0) + 1;
                    }
                }
                else if (parentLocal && !parentLocal.isUnsafe) {
                    local.isUnsafe = false;
                }
                if (parentLocal === null || parentLocal === void 0 ? void 0 : parentLocal.isIterator) {
                    local.isIterator = parentLocal.isIterator;
                }
                if (!local.isUsed && isLoop) {
                    // avoid false positive if a local set in a loop isn't used
                    const someParentLocal = findLocal(local.name);
                    if (someParentLocal === null || someParentLocal === void 0 ? void 0 : someParentLocal.isUsed) {
                        local.isUsed = true;
                    }
                }
                parent.locals.set(name, local);
            });
        }
    }
    function visitExpression(expr, parent, curr) {
        if (brighterscript_1.isVariableExpression(expr)) {
            const name = expr.name.text;
            if (name === 'm') {
                return;
            }
            const local = findLocal(name);
            if (!local) {
                deferred.push({
                    kind: expr.isCalled ? ValidationKind.UninitialisedFn : ValidationKind.UninitializedVar,
                    name: name,
                    range: expr.range
                });
                return;
            }
            else {
                local.isUsed = true;
                verifyVarCasing(local, expr.name);
            }
            if (local.isUnsafe && !findSafeLocal(name)) {
                if (local.isIterator) {
                    diagnostics.push({
                        severity: severity.unsafeIterators,
                        code: VarLintError.UnsafeIteratorVar,
                        message: `Using iterator variable '${name}' outside loop`,
                        range: expr.range,
                        file: file
                    });
                }
                else if (!isNarrowing(local, expr, parent, curr)) {
                    diagnostics.push({
                        severity: severity.unsafePathLoop,
                        code: VarLintError.UnsafeInitialization,
                        message: `Not all the code paths assign '${name}'`,
                        range: expr.range,
                        file: file
                    });
                }
            }
        }
    }
    function isNarrowing(local, expr, parent, curr) {
        var _a;
        // Are we inside an `if/elseif` statement condition?
        if (!brighterscript_1.isIfStatement(curr.stat)) {
            return false;
        }
        const block = curr.stat.thenBranch;
        // look for a statement testing whether variable is `invalid`,
        // like `if x <> invalid` or `else if x = invalid`
        if (!brighterscript_1.isBinaryExpression(parent) || !(brighterscript_1.isLiteralInvalid(parent.left) || brighterscript_1.isLiteralInvalid(parent.right))) {
            // maybe the variable was narrowed as part of the condition
            // e.g. 2nd condition in: if x <> invalid and x.y = z
            return (_a = curr.narrows) === null || _a === void 0 ? void 0 : _a.some(narrow => narrow.text === local.name);
        }
        const operator = parent.operator.kind;
        if (operator !== brighterscript_1.TokenKind.Equal && operator !== brighterscript_1.TokenKind.LessGreater) {
            return false;
        }
        const narrow = {
            text: local.name,
            range: local.range,
            type: operator === brighterscript_1.TokenKind.Equal ? 'invalid' : 'valid',
            block
        };
        if (!curr.narrows) {
            curr.narrows = [];
        }
        curr.narrows.push(narrow);
        return true;
    }
    function finalize(locals) {
        locals.forEach(local => {
            if (!local.isUsed && !local.isIterator) {
                diagnostics.push({
                    severity: severity.unusedVariable,
                    code: VarLintError.UnusedVariable,
                    message: `Variable '${local.name}' is set but value is never used`,
                    range: local.range,
                    file: file
                });
            }
        });
    }
    return {
        openBlock: openBlock,
        closeBlock: closeBlock,
        visitStatement: visitStatement,
        visitExpression: visitExpression
    };
}
exports.createVarLinter = createVarLinter;
function runDeferredValidation(lintContext, scope, files, callables) {
    const globals = lintContext.globals;
    const diagnostics = [];
    files.forEach((file) => {
        const deferred = deferredValidation.get(file.pathAbsolute);
        if (deferred) {
            deferredVarLinter(scope, file, callables, globals, deferred, diagnostics);
        }
    });
    return diagnostics;
}
exports.runDeferredValidation = runDeferredValidation;
function deferredVarLinter(scope, file, callables, globals, deferred, diagnostics) {
    // lookups for namespaces and classes
    const toplevel = new Set(globals);
    scope.getAllNamespaceStatements().forEach(ns => {
        toplevel.add(ns.name.toLowerCase().split('.')[0]); // keep root of namespace
    });
    scope.getClassMap().forEach(cls => {
        const name = cls.item.name.text.toLowerCase();
        if (!cls.item.namespaceName) {
            toplevel.add(name);
        }
    });
    if (brighterscript_1.isBrsFile(file)) {
        file.parser.references.classStatements.forEach(cls => {
            toplevel.add(cls.name.text.toLowerCase());
        });
    }
    deferred.forEach(({ kind, name, local, range }) => {
        const key = name === null || name === void 0 ? void 0 : name.toLowerCase();
        const hasCallable = key ? callables.has(key) || toplevel.has(key) : false;
        switch (kind) {
            case ValidationKind.UninitializedVar:
                if (!hasCallable) {
                    diagnostics.push({
                        severity: brighterscript_1.DiagnosticSeverity.Error,
                        code: VarLintError.UninitializedVar,
                        message: `Using uninitialised variable '${name}' when this file is included in scope '${scope.name}'`,
                        range: range,
                        file: file
                    });
                }
                // TODO else test case
                break;
            case ValidationKind.UninitialisedFn:
                if (!hasCallable) {
                    diagnostics.push({
                        severity: brighterscript_1.DiagnosticSeverity.Error,
                        code: VarLintError.UninitializedVar,
                        message: `Using uninitialised variable '${name}' when this file is included in scope '${scope.name}'`,
                        range: range,
                        file: file
                    });
                }
                // TODO else test case
                break;
            case ValidationKind.Assignment:
                break;
            case ValidationKind.Unsafe:
                break;
        }
    });
}
